README.md
# Montclair Transmogrifier

This project takes a base version of montclair, montclair-pwa-android,
and/or montclair-pwa-ios and generates a white label build. You should
not run this project directly, but instead, it should be included in a
whitelabel-assets project.

## Info

Copyright (c) Marcus Dillavou <line72@line72.net>
Released under the GPLv3


ChangeLog
1.2.1
==========
2023-08-23  Marcus Dillavou  <line72@line72.net>
	* Fix version in config.xml

1.2.0
==========
2023-08-23  Marcus Dillavou  <line72@line72.net>
	* ISSUE-3 - Update Android Runner for new Cordova

1.0.0
==========
2020-01-24  Marcus Dillavou  <line72@line72.net>
	* Updated the web runner to support the 1.6.0 code changes

0.9.6
==========
2020-01-19  Marcus Dillavou  <line72@line72.net>
	* Update the meta description

0.9.5
==========
2019-12-30  Marcus Dillavou  <line72@line72.net>
	* Correctly update the version in the Android Manifest

0.9.4
==========
2019-12-29  Marcus Dillavou  <line72@line72.net>
	* Updated icons and launch images for iphonex

0.9.3
==========
2019-12-29  Marcus Dillavou  <line72@line72.net>
	* Fixed issue in iOS parsing the config.xml

0.9.2
==========
2019-12-29  Marcus Dillavou  <line72@line72.net>
	* Updated the iOS runner for the new project

0.9.1
==========
2019-12-20  Marcus Dillavou  <line72@line72.net>
	* ISSUE-1 - Support new favicons in 1.2.1

0.9.0
==========
2019-12-19  Marcus Dillavou  <line72@line72.net>
	* Initial version



config.py

class Config:
    def __init__(self,
                 build_dir = './build', # base location of the builds
                 repo = 'whitelabel', # The repo suffix
                 package_name = 'montclair-whitelabel', # App package name
                 name = 'Montclair Whitelabel', # Project name, this shows up in multiple places
                 description = 'Real time Bus Tracker',
                 url = 'https://whitelabel.montclair.line72.net',
                 logo_svg = '', # Path to the logo.svg (used to generate favicon.ico and app-icon.png)
                 montclair_config = None,
                 ios_config = None,
                 android_config = None
                 ):
        self.build_dir = build_dir
        self.repo = repo
        self.package_name = package_name
        self.name = name
        self.description = description
        self.url = url
        self.logo_svg = logo_svg
        self.montclair_config = montclair_config
        self.ios_config = ios_config
        self.android_config = android_config

class MontclairConfig:
    def __init__(self,
                 version = '1.0.0', # version of montclair to build from
                 revision = 1, # output revision of whitelabel build,
                 title = None, # optional title
                 first_run_text = '', # text to show the first time the app is run
                 configuration_js_file = '' # path to the Configuration.js
    ):
        self.version = version
        self.revision = revision
        self.title = title
        self.first_run_text = first_run_text
        self.configuration_js_file = configuration_js_file

class MontclairiOSConfig:
    def __init__(self,
                 version = '1.0.0', # version of montclair-pwa-ios to build from
                 revision = 1, # output revision of whitelabel build
                 app_id = 'net.line72.montclair.whitelabel',
                 app_store_id = '',
                 app_store_url = '' # URL to the app store
    ):
        self.version = version
        self.revision = revision
        self.app_id = app_id
        self.app_store_id = app_store_id
        self.app_store_url = app_store_url

class MontclairAndroidConfig:
    def __init__(self,
                 version = '1.0.0', # version of montclair-pwa-android to build from
                 revision = 1, # output revision of whitelabel build
                 app_id = 'net.line72.montclair.whitelabel',
                 play_store_url = '' # URL to the play store
    ):
        self.version = version
        self.revision = revision
        self.app_id = app_id
        self.play_store_url = play_store_url


runner.py
import sys
import argparse

from transmogrifier.runners.web import Web
from transmogrifier.runners.ios import IOS
from transmogrifier.runners.android import Android

class Runner:
    def __init__(self, config):
        self.config = config
        self.web = Web(config)
        self.ios = IOS(config)
        self.android = Android(config)

    def go(self):
        self.parse_args()
        
        print('running')
        self.web.go()
        self.ios.go()
        self.android.go()
        
    def parse_args(self):
        parser = argparse.ArgumentParser()
        parser.add_argument('-e', '--env', action = 'store_true',
                            help = 'Write out an ENV file')

        args = parser.parse_args()
        if args.env:
            self.write_env()
            sys.exit(0)
        
    def write_env(self):
        with open('ENV', 'w') as f:
            f.write(f'export BUILD_DIR="{self.config.build_dir}"\n')
            f.write(f'export REPO="{self.config.repo}"\n')
            f.write(f'export VERSION="{self.config.montclair_config.version}"\n')
            f.write(f'export OUTPUT_VERSION="{self.config.montclair_config.version}-{self.config.montclair_config.revision}"\n')
            
            f.write(f'export HAS_IOS={1 if self.config.ios_config else 0}\n')
            if self.config.ios_config:
                f.write(f'export IOS_VERSION="{self.config.ios_config.version}"\n')
                f.write(f'export OUTPUT_IOS_VERSION="{self.config.ios_config.version}-{self.config.ios_config.revision}"\n')
                
            f.write(f'export HAS_ANDROID={1 if self.config.android_config else 0}\n')
            if self.config.android_config:
                f.write(f'export ANDROID_VERSION="{self.config.android_config.version}"\n')
                f.write(f'export OUTPUT_ANDROID_VERSION="{self.config.android_config.version}-{self.config.android_config.revision}"\n')


runners/ios.py
# -*- mode: python -*-
import sys
import os
import subprocess
import xml.etree.ElementTree as ET
import urllib.parse
import json

class IOS:
    def __init__(self, config):
        self.config = config

    def go(self):
        if self.config.ios_config is None:
            print('Skipping iOS PWA creation...', file = sys.stderr)
            return

        self.update_project_pbx()
        self.update_xcscheme()
        self.update_package_json()
        self.update_config_xml()
        self.update_plist()
        self.update_manifest()
        self.update_ios_json()
        self.update_index()
        self.update_generation_info()
        self.create_icons()
        self.create_splash_screen()

    def update_project_pbx(self):
        # update platforms/ios/Montclair.xcodeproj/project.pbxproj
        #!mwd - Nothing to do...
        pass

    def update_xcscheme(self):
        pass
    
    def update_package_json(self):
        # update the package.json and the package-lock.json
        package = json.loads(self.oread('package.json'))
        # replace the name
        package['name'] = self.config.package_name
        package['displayName'] = self.config.name
        
        with self.o('package.json', 'w') as f:
            f.write(json.dumps(package, indent = 2))
            f.write('\n')

        package_lock = json.loads(self.oread('package-lock.json'))
        # replace the name
        package_lock['name'] = self.config.package_name
        
        with self.o('package-lock.json', 'w') as f:
            f.write(json.dumps(package_lock, indent = 2))
            f.write('\n')

    def update_config_xml(self):
        # update config.xml and platforms/ios/Montclair/config.xml
        for i in ('config.xml', 'platforms/ios/Montclair/config.xml'):
            tree = ET.parse(self.base_path(i))
            root = tree.getroot()
        
            # register the namespace
            # See: https://stackoverflow.com/questions/3895951/create-svg-xml-document-without-ns0-namespace-using-python-elementtree
            ns = 'http://www.w3.org/ns/widgets'
            ET.register_namespace('', ns)
            ET.register_namespace('cdv', 'http://cordova.apache.org/ns/1.0')
            
            # update the root id and version
            root.set('id', self.config.ios_config.app_id)
            root.set('version', self.config.montclair_config.version)
            # find the name, description, and content nodes, allow-navigation and update them
            name = root.find(f'.//{{{ns}}}name')
            name.text = self.config.name
            
            description = root.find(f'.//{{{ns}}}description')
            description.text = self.config.description
            
            content = root.find(f'.//{{{ns}}}content')
            content.set('src', f'{self.config.url}/index.html')
            
            nav = root.find(f'.//{{{ns}}}allow-navigation')
            if nav != None:
                nav.set('href', f'{self.config.url}/*')
            
            # write back out
            tree.write(self.base_path(i),
                       encoding='utf-8', xml_declaration = True,
                       default_namespace = '')
            
    def update_plist(self):
        # Update platforms/ios/Montclair/Montclair-Info.plist

        url = urllib.parse.urlparse(self.config.url).netloc
        
        plist = self.oread('platforms/ios/Montclair/Montclair-Info.plist')
        plist = plist.replace(
            'net.line72.montclair', self.config.ios_config.app_id
        ).replace(
            'montclair.line72.net', url
        ).replace(
            '<string>Montclair</string>', f'<string>{self.config.name}</string>'
        )

        with self.o('platforms/ios/Montclair/Montclair-Info.plist', 'w') as f:
            f.write(plist)

    def update_manifest(self):
        # Update manifest.json AND www/manifest.json, platforms/ios/www/manifest.json
        for i in ('manifest.json', 'www/manifest.json', 'platforms/ios/www/manifest.json'):
            m = json.loads(self.oread(i))
            m['short_name'] = self.config.name
            m['name'] = self.config.name
            m['description'] = self.config.description
            m['start_url'] = self.config.url + '/index.html'

            # fix all the icons
            for icon in m['icons']:
                url = urllib.parse.urlparse(icon['src'])
                icon['src'] = f'{self.config.url}{url.path}'
            
            with self.o(i, 'w') as f:
                f.write(json.dumps(m, indent = 4))
                f.write('\n')

    def update_ios_json(self):
        # update platforms/ios/ios.json and plugins/ios.json
        for i in ('platforms/ios/ios.json', 'plugins/ios.json'):
            r = self.oread(i)
            r = r.replace('net.line72.montclair', self.config.ios_config.app_id)

            with self.o(i, 'w') as f:
                f.write(r)

    def update_index(self):
        # Update www/index.html
        i = self.oread('www/index.html')
        with self.o('www/index.html', 'w') as f:
            f.write(i.replace('Montclair', self.config.name))

    def update_generation_info(self):
        # Update generationInfo.json
        for i in ('generationInfo.json', 'platforms/ios/generationInfo.json'):
            g = json.loads(self.oread(i))
            g['generatedURL'] = f'{self.config.url}/manifest.json'
            with self.o(i, 'w') as f:
                f.write(json.dumps(g, indent = 4))
                f.write('\n')

    def create_icons(self):
        # app-icon.png 512x512
        subprocess.run(['convert',
                        '-size', '512x512',
                        'xc:none',
                        '-fill', 'white',
                        '-draw', 'roundRectangle 0,0 512,512 50,50',
                        self.config.logo_svg,
                        '-resize', '512x512',
                        '-compose', 'SrcIn',
                        '-composite',
                        self.base_path('app-icon.png')],
                       check = True)

        # favicon.ico with multiple resolutions
        subprocess.run(['convert',
                        '-size', '256x256',
                        'xc:none',
                        '-fill', 'white',
                        '-draw', 'roundRectangle 0,0 256,256 25,25',
                        self.config.logo_svg,
                        '-resize', '256x256',
                        '-compose', 'SrcIn',
                        '-composite',
                        '-define', 'icon:auto-resize=256,192,152,144,128,96,72,64,48,32,24,16',
                        self.base_path('favicon.ico')],
                       check = True)

        # create a bunch of icons with transparency
        icons = [
            ('AppIcon29x29@2x.png', '58x58'),
            ('AppIcon40x40@2x.png', '80x80'),
            ('icon.png', '57x57'),
            ('icon120-1.png', '120x120'),
            ('icon120.png', '120x120'),
            ('icon152.png', '152x152'),
            ('icon167.png', '167x167'),
            ('icon180.png', '180x180'),
            ('icon20.png', '20x20'),
            ('icon29-1.png', '29x29'),
            ('icon29.png', '29x29'),
            ('icon40-1.png', '40x40'),
            ('icon40-2.png', '40x40'),
            ('icon40.png', '40x40'),
            ('icon58-1.png', '58x58'),
            ('icon58.png', '58x58'),
            ('icon60.png', '60x60'),
            ('icon76.png', '76x76'),
            ('icon80-1.png', '80x80'),
            ('icon80.png', '80x80'),
            ('icon87.png', '87x87'),
            ('icon-20.png', '20x20'),
            ('icon-20@2x.png', '40x40'),
            ('icon-20@3x.png', '60x60'),
            ('icon-40.png', '40x40'),
            ('icon-40@2x.png', '80x80'),
            ('icon-50.png', '50x50'),
            ('icon-50@2x.png', '100x100'),
            ('icon-60@2x.png', '120x120'),
            ('icon-60@3x.png', '180x180'),
            ('icon-72.png', '72x72'),
            ('icon-72@2x.png', '144x144'),
            ('icon-76.png', '76x76'),
            ('icon-76@2x.png', '152x152'),
            ('icon-83.5@2x.png', '167x167'),
            ('icon@2x.png', '114x114'),
            ('icon-small.png', '29x29'),
            ('icon-small@2x.png', '58x58'),
            ('icon-small@3x.png', '87x87'),
            ('icon-1024.png', '1024x1024')
        ]

        for i in icons:
            subprocess.run(['convert', self.config.logo_svg,
                            '-resize', i[1],
                            self.base_path(f'platforms/ios/Montclair/Images.xcassets/AppIcon.appiconset/{i[0]}')],
                           check = True)

        
        # create icons without transparency
        subprocess.run(['convert', self.config.logo_svg,
                        '-alpha', 'off',
                        '-resize', '1024x1024',
                        self.base_path(f'platforms/ios/Montclair/Images.xcassets/AppIcon.appiconset/icon1024-no-transparency.png')],
                       check = True)


    def create_splash_screen(self):
        icons = [
            ('Default-568h@2x~iphone.png', '640x1136'),
            ('Default-667h.png', '750x1134'),
            ('Default-736h.png', '1242x2208'),
            ('Default@2x~iphone.png', '640x960'),
            ('Default~iphone.png', '320x480'),
            ('Default-Landscape-736h.png', '2208x1242'),
            ('Default-Landscape@2x~ipad.png', '2048x1536'),
            ('Default-Landscape~ipad.png', '1024x768'),
            ('Default-Portrait@2x~ipad.png', '1536x2048'),
            ('Default-Portrait~ipad.png', '768x1024'),
            ('Default-2436h.png', '1125x2436'),
            ('Default-Landscape-2436h.png', '2436x1125'),
            ('launch_image1024x748.png', '1024x768'),
            ('launch_image1024x768-1.png', '1024x768'),
            ('launch_image1024x768.png', '1024x768'),
            ('launch_image1242x2208.png', '1242x2208'),
            ('launch_image1536x2008.png', '1536x2008'),
            ('launch_image1536x2048-1.png', '1536x2048'),
            ('launch_image1536x2048.png', '1536x2048'),
            ('launch_image2048x1496.png', '2048x1496'),
            ('launch_image2048x1536-1.png', '2048x1536'),
            ('launch_image2048x1536.png', '2048x1536'),
            ('launch_image2208x1242.png', '2208x1242'),
            ('launch_image320x480-9.png', '320x480'),
            ('launch_image640x1136-1.png', '640x1136'),
            ('launch_image640x1136.png', '640x1136'),
            ('launch_image640x960-1.png', '640x960'),
            ('launch_image640x960.png', '640x960'),
            ('launch_image750x1334.png', '750x1334'),
            ('launch_image768x1004.png', '768x1004'),
            ('launch_image768x1024-1.png', '768x1024'),
            ('launch_image768x1024.png', '768x1024')
        ]
        for i in icons:
            w, h = [int(x) for x in i[1].split('x')]
            m = min(w, h)

            # our logo is sqare, so this in the new resolution
            img = int(m / 4)

            # compute the frame, divide by two, since it is on both sides
            frame_w = (w - img) / 2
            frame_h = (h - img) / 2
            
            subprocess.run(['convert', self.config.logo_svg,
                            '-alpha', 'off',
                            '-resize', f'{img}x{img}',
                            '-mattecolor', 'White',
                            '-frame', f'{frame_w}x{frame_h}',
                            self.base_path(f'platforms/ios/Montclair/Images.xcassets/LaunchImage.launchimage/{i[0]}')],
                           check = True)

    def base_path(self, fname):
        return os.path.join(self.config.build_dir, f'montclair-{self.config.repo}-pwa-ios', fname)

    def o(self, fname, mode = 'r'):
        return open(self.base_path(fname), mode)

    def oread(self, fname):
        with self.o(fname) as f:
            return f.read()

    def oreadlines(self, fname):
        with self.o(fname) as f:
            return f.readlines()


runners/android.py
# -*- mode: python -*-

import sys
import os
import subprocess
import xml.etree.ElementTree as ET
import urllib.parse
import json

class Android:
    def __init__(self, config):
        self.config = config

    def go(self):
        if self.config.android_config is None:
            print('Skipping Android PWA creation...', file = sys.stderr)
            return

        self.update_android_json()
        self.update_config_xml()
        self.update_strings_xml()
        self.update_manifest_json()
        self.update_android_manifest()
        self.update_package_name()
        self.update_generation_info()
        self.create_icons()

    def update_android_json(self):
        # Update platforms/android/android.json and plugins/android.json
        for i in ('platforms/android/android.json', 'plugins/android.json'):
            m = self.oread(i)
            m = m.replace('net.line72.montclair', self.config.android_config.app_id)

            with self.o(i, 'w') as f:
                f.write(m)

    def update_config_xml(self):
        # Updates platforms/android/res/xml/config.xml and config.xml
        for i in ('config.xml', 'platforms/android/app/src/main/res/xml/config.xml'):
            tree = ET.parse(self.base_path(i))
            root = tree.getroot()
        
            # register the namespace
            # See: https://stackoverflow.com/questions/3895951/create-svg-xml-document-without-ns0-namespace-using-python-elementtree
            ns = 'http://www.w3.org/ns/widgets'
            ET.register_namespace('', ns)
            ET.register_namespace('cdv', 'http://cordova.apache.org/ns/1.0')
            
            # update the root id and version
            root.set('id', self.config.android_config.app_id)
            root.set('version', self.config.android_config.version)
            # find the name, description, and content nodes, allow-navigation and update them
            name = root.find(f'.//{{{ns}}}name')
            name.text = self.config.name
            
            description = root.find(f'.//{{{ns}}}description')
            description.text = self.config.description
            
            content = root.find(f'.//{{{ns}}}content')
            content.set('src', f'{self.config.url}/index.html')
            
            nav = root.find(f'.//{{{ns}}}allow-navigation')
            nav.set('href', f'{self.config.url}/*')
            
            # write back out
            tree.write(self.base_path(i),
                       encoding='utf-8', xml_declaration = True,
                       default_namespace = '')

    def update_strings_xml(self):
        # Update platforms/android/res/values/strings.xml
        p = 'platforms/android/app/src/main/res/values/strings.xml'
        
        tree = ET.parse(self.base_path(p))
        root = tree.getroot()

        name = root.find(f".//string[@name='app_name']")
        name.text = self.config.name
        
        # write back out
        tree.write(self.base_path(p),
                   encoding='utf-8', xml_declaration = True,
                   default_namespace = '')

    def update_manifest_json(self):
        # Update ./platforms/android/assets/www/manifest.json, www/manifest.json, and manifest.json
        for i in ('manifest.json', 'www/manifest.json', 'platforms/android/app/src/main/assets/www/manifest.json'):
            m = json.loads(self.oread(i))
            m['short_name'] = self.config.name
            m['name'] = self.config.name
            m['description'] = self.config.description
            m['start_url'] = self.config.url + '/index.html'

            # fix all the icons
            for icon in m['icons']:
                url = urllib.parse.urlparse(icon['src'])
                icon['src'] = f'{self.config.url}{url.path}'
            
            with self.o(i, 'w') as f:
                f.write(json.dumps(m, indent = 4))
                f.write('\n')

    def update_android_manifest(self):
        # Update ./platforms/android/AndroidManifest.xml
        p = 'platforms/android/app/src/main/AndroidManifest.xml'

        tree = ET.parse(self.base_path(p))
        root = tree.getroot()

        ns = 'http://schemas.android.com/apk/res/android'
        ET.register_namespace('android', ns)

        root.set(f'{{{ns}}}versionName', f'{self.config.android_config.version}-{self.config.android_config.revision}')
        root.set('package', self.config.android_config.app_id)

        # write back out
        tree.write(self.base_path(p),
                   encoding='utf-8', xml_declaration = True,
                   default_namespace = '')

    def update_package_name(self):
        # Update the package net.line72.net.montclair in all the java files
        src_dir = os.path.join('platforms', 'android', 'app', 'src', 'main', 'java')

        # First make the new directory based on the package name
        package_path = os.path.join(*self.config.android_config.app_id.split('.'))
        full_path = self.base_path(os.path.join(src_dir, package_path))
        os.makedirs(full_path)
        # Next move move the net/line72/montclair/MainActivity.java to our new package name directory.
        old_path = self.base_path(os.path.join(src_dir, 'net', 'line72', 'montclair'))
        os.rename(os.path.join(old_path, 'MainActivity.java'), os.path.join(full_path, 'MainActivity.java'))
    
        # Then update the package name in:
        # ./platforms/android/src/main/java/net/line72/NEW_PACKAGE_NAME/MainActivity.java
        fname = os.path.join(src_dir, package_path, 'MainActivity.java')
        j = self.oread(fname)

        j = j.replace('net.line72.montclair', self.config.android_config.app_id)
        with self.o(fname, 'w') as f:
            f.write(j)

    def update_generation_info(self):
        # update ./generationInfo.json
        g = json.loads(self.oread('generationInfo.json'))
        g['generatedURL'] = f'{self.config.url}/manifest.json'
        with self.o('generationInfo.json', 'w') as f:
            f.write(json.dumps(g, indent = 4))
            f.write('\n')

    def create_icons(self):
        # app-icon.png 512x512
        subprocess.run(['convert',
                        '-size', '512x512',
                        'xc:none',
                        '-fill', 'white',
                        '-draw', 'roundRectangle 0,0 512,512 50,50',
                        self.config.logo_svg,
                        '-resize', '512x512',
                        '-compose', 'SrcIn',
                        '-composite',
                        self.base_path('app-icon.png')],
                       check = True)

        # favicon.ico with multiple resolutions
        subprocess.run(['convert',
                        '-size', '256x256',
                        'xc:none',
                        '-fill', 'white',
                        '-draw', 'roundRectangle 0,0 256,256 25,25',
                        self.config.logo_svg,
                        '-resize', '256x256',
                        '-compose', 'SrcIn',
                        '-composite',
                        '-define', 'icon:auto-resize=256,192,152,144,128,96,72,64,48,32,24,16',
                        self.base_path('favicon.ico')],
                       check = True)

        
        icons = [
            ('mipmap-xxxhdpi/ic_launcher.png', '192x192'),
            ('mipmap-xxxhdpi-v26/ic_launcher_monochrome.png', '432x432'),
            ('mipmap-xxxhdpi-v26/ic_launcher_foreground.png', '432x432'),
            ('mipmap-xxhdpi/ic_launcher.png', '144x144'),
            ('mipmap-xxhdpi-v26/ic_launcher_monochrome.png', '324x324'),
            ('mipmap-xxhdpi-v26/ic_launcher_foreground.png', '324x324'),
            ('mipmap-xhdpi/ic_launcher.png', '96x96'),
            ('mipmap-xhdpi-v26/ic_launcher_monochrome.png', '216x216'),
            ('mipmap-xhdpi-v26/ic_launcher_foreground.png', '216x216'),
            ('mipmap-hdpi/ic_launcher.png', '72x72'),
            ('mipmap-hdpi-v26/ic_launcher_monochrome.png', '163x163'),
            ('mipmap-hdpi-v26/ic_launcher_foreground.png', '163x163'),
            ('mipmap-mdpi/ic_launcher.png', '48x48'),
            ('mipmap-mdpi-v26/ic_launcher_monochrome.png', '108x108'),
            ('mipmap-mdpi-v26/ic_launcher_foreground.png', '108x108'),
            ('mipmap-ldpi/ic_launcher.png', '36x36'),
            ('mipmap-ldpi-v26/ic_launcher_monochrome.png', '36x36'),
            ('mipmap-ldpi-v26/ic_launcher_foreground.png', '36x36')
        ]
        for i in icons:
            c = i[1].split('x')
            corners = ','.join(c)
            corner_ratio = 50 / 512
            corner_size = int(int(c[0]) * corner_ratio)
            subprocess.run(['convert',
                            '-size', i[1],
                            'xc:none',
                            '-fill', 'white',
                            '-draw', f'roundRectangle 0,0 {corners} {corner_size},{corner_size}',
                            self.config.logo_svg,
                            '-resize', i[1],
                            '-compose', 'SrcIn',
                            '-composite',
                            self.base_path(f'platforms/android/app/src/main/res/{i[0]}')],
                           check = True)
    
    def base_path(self, fname):
        return os.path.join(self.config.build_dir, f'montclair-{self.config.repo}-pwa-android', fname)

    def o(self, fname, mode = 'r'):
        return open(self.base_path(fname), mode)

    def oread(self, fname):
        with self.o(fname) as f:
            return f.read()

    def oreadlines(self, fname):
        with self.o(fname) as f:
            return f.readlines()
    
